compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/JvmInlineClassLowering.kt

```diff
@@ -679,35 +679,64 @@ private class JvmInlineClassLowering(context: JvmBackendContext) : JvmValueClass
+            function.body = irBlockBody {
+                val left = irTemporary(
+                    coerceInlineClasses(
+                        irGet(function.valueParameters[0]),
+                        function.valueParameters[0].type,
+                        context.irBuiltIns.anyType
+                    )
                 )
-            } + inlineSubclasses.map {
-                val eq = this@JvmInlineClassLowering.context.inlineClassReplacements
-                    .getSpecializedEqualsMethod(it.owner, context.irBuiltIns)
-                irBranch(
-                    irCallOp(
-                        boolAnd, context.irBuiltIns.booleanType,
-                        irIs(irGet(left), getInlineClassUnderlyingType(it.owner)),
-                        irIs(irGet(right), getInlineClassUnderlyingType(it.owner)),
-                    ),
-                    irCall(eq).apply {
-                        putValueArgument(0, irGet(left))
-                        putValueArgument(1, irGet(right))
-                    }
+                val right = irTemporary(
+                    coerceInlineClasses(
+                        irGet(function.valueParameters[1]),
+                        function.valueParameters[1].type,
+                        context.irBuiltIns.anyType
+                    )
                 )
-            } + irBranch(irTrue(), irFalse())
-            irExprBody(irWhen(context.irBuiltIns.booleanType, branches))
+                val branches = noinlineSubclasses.map {
+                    irBranch(
+                        irCallOp(
+                            boolAnd, context.irBuiltIns.booleanType,
+                            irIs(irGet(left), it.owner.defaultType),
+                            irIs(irGet(right), it.owner.defaultType),
+                        ),
+                        irReturn(irCallOp(equals, context.irBuiltIns.booleanType, irGet(left), irGet(right)))
+                    )
+                } + inlineSubclasses.map {
+                    val eq = this@JvmInlineClassLowering.context.inlineClassReplacements
+                        .getSpecializedEqualsMethod(it.owner, context.irBuiltIns)
+                    val underlyingType = getInlineClassUnderlyingType(it.owner)
+                    irBranch(
+                        irCallOp(
+                            boolAnd, context.irBuiltIns.booleanType,
+                            irIs(irGet(left), underlyingType),
+                            irIs(irGet(right), underlyingType),
+                        ),
+                        irReturn(
+                            irCall(eq).apply {
+                                putValueArgument(
+                                    0, coerceInlineClasses(
+                                        irImplicitCast(irGet(left), underlyingType),
+                                        underlyingType,
+                                        it.owner.defaultType
+                                    )
```

Comment:
Is this necessary? Couldn't you cast `function.valueParameters[0], ...[1]` instead?